Readme.md
# Haml.js

  High performance JavaScript [Haml](http://haml-lang.com) implementation for [nodejs](http://nodejs.org)
  
  For a higher quality implementation you may want to look at my [Jade](http://jade-lang.com) template engine,
  however the syntax is slightly different. Jade's engine may be back-ported to haml.js in the future.

## Installation

      $ npm install hamljs

      node> require('hamljs')

## Express

 To use with [Express](http://expressjs.com) and the .haml extension, simply register the engine:
 
     app.engine('.haml', require('hamljs').renderFile);

## About

  Benchmarks rendering the same 21 line haml file located at _benchmarks/page.haml_,
  shows that this library is nearly **65%** or **3 times** faster than haml-js.
  
      Winner: haml.js
      Compared with next highest (haml-js), it's:
      65.39% faster
      2.89 times as fast
      0 order(s) of magnitude faster

  Haml.js attempts to comply with the original [Haml](http://haml-lang.com/docs/yardoc/file.HAML_REFERENCE.html)
  implementation as well as possible. There are no magic "plugins" like
  found in other JavaScript haml implementations, for example the following
  will work just fine:
  
    - if (items)
      %ul
        - for (var i = 0; i < items.length; ++i)
          %li= items[i]
  
  Iteration is the one exception to these magical plugins,
  since this is **ugly** in JavaScript, you may also:
  
    - if (items)
      %ul
        - each item in items
          %li= item
          
## Tags

    %div text

html:

    <div>text</div>
    
## Classes

    %div.article.first
      article text here
      and here
      
html:

    <div class="article first">
      article text here and here
    </div>
    
## Div Class Shortcut

    .comment hey
    
html:

    <div class="comment">hey</div>
    
## Div Id Shortcut

    #article-1 foo
    
html:

    <div id="article-1">foo</div>
    
## Combining Ids and Classes

You may chain id and classes in any order:

    .article#first.summary content
    
html:

    <div id="first" class="article summary">context</div>
    
## Attributes

    %a{ href: 'http://google.com', title: 'Google It' } Google
    
html:

    <a href="http://google.com" title="Google It">Google</a>

Attribute keys such as "for" are automatically quoted
by haml.js, so instead of:

    %label{ 'for': 'something' }

you should:

    %label{ for: 'something' }

which will render:

    <label for="something"></label>
    
## Boolean Attributes

    %input{ type: 'checkbox', checked: true }
    
html:

    <input type="checkbox" checked="checked"/>
    
## Combining Attributes, Ids, and Classes
    
Wemay also contain id and classes before or after:

    %a.button{ href: 'http://google.com', title: 'Google It' }.first Google
    
html:

    <a href="http://google.com" title="Google It" class="button first">Google</a>
    
## Code

Code starting with a hyphen will be executed but
not buffered, where as code using the equals sign
will be buffered:

    - a = 1
    - b = 2
    = a + b
    
html:

    3
    
HTML buffered with equals sign will **always** be escaped:

    = "<br/>"
    
html:
   
    &lt;br/&gt;
    
To prevent escaping of HTML entities we can use _!=_:

    != "<br/>"
    
html:

    <br/>
    
## Iteration

    %ul
      - each item in items
        %li= item
        
html:

    <ul>
      <li>one</li>
      <li>two</li>
      <li>three</li>
    </ul>
    
If you require the key or index of the object
or array during iteration simple append a comma
following another id:

    %ul
      - each item, index in items
        %li= item + '(' + index + ')'
        
html:

    <ul>
      <li>one(0)</li>
      <li>two(1)</li>
      <li>three(2)</li>
    </ul>
    
## Doctypes

Defaults to transitional:
    
    !!!
    
html:

    <!DOCTYPE html PUBLIC "-//W3C//DTD XHTML 1.0 Transitional//EN" "http://www.w3.org/TR/xhtml1/DTD/xhtml1-transitional.dtd">
    
Optionally pass a supported doctype name:

    !!! strict
    
html:

    <!DOCTYPE html PUBLIC "-//W3C//DTD XHTML 1.0 Strict//EN" "http://www.w3.org/TR/xhtml1/DTD/xhtml1-strict.dtd">

currently supported doctypes, which can be
extended simply by adding values to to _haml.doctypes_.

    '5': '<!DOCTYPE html>',
    'xml': '<?xml version="1.0" encoding="utf-8" ?>',
    'default': '<!DOCTYPE html PUBLIC "-//W3C//DTD XHTML 1.0 Transitional//EN" "http://www.w3.org/TR/xhtml1/DTD/xhtml1-transitional.dtd">',
    'strict': '<!DOCTYPE html PUBLIC "-//W3C//DTD XHTML 1.0 Strict//EN" "http://www.w3.org/TR/xhtml1/DTD/xhtml1-strict.dtd">',
    'frameset': '<!DOCTYPE html PUBLIC "-//W3C//DTD XHTML 1.0 Frameset//EN" "http://www.w3.org/TR/xhtml1/DTD/xhtml1-frameset.dtd">',
    '1.1': '<!DOCTYPE html PUBLIC "-//W3C//DTD XHTML 1.1//EN" "http://www.w3.org/TR/xhtml11/DTD/xhtml11.dtd">',
    'basic': '<!DOCTYPE html PUBLIC "-//W3C//DTD XHTML Basic 1.1//EN" "http://www.w3.org/TR/xhtml-basic/xhtml-basic11.dtd">',
    'mobile': '<!DOCTYPE html PUBLIC "-//WAPFORUM//DTD XHTML Mobile 1.2//EN" "http://www.openmobilealliance.org/tech/DTD/xhtml-mobile12.dtd">'


## :cdata

    %script
      :cdata
        foo
        
html:

    <script><![CDATA[
    foo
    ]]></script>
    
## :javascript

    %head
      :javascript
        if (foo)
          if (bar)
            alert('baz')
      
html:

    <head>
      <script type="javascript">
      //<![CDATA[
      if (foo)
        if (bar)
          alert('baz')
      //]]>
      </script>
    </head>
    
## Extending Haml

### Adding Filters

    var haml = require('haml')
    haml.filters.my_filter = function(str) {
      return doSomethingWith(str)
    }

by registering the filter function _my_filter_ we can now
utilize it within our Haml templates as shown below:
    %p
      :my_filter
        some text
        here yay
        whoop awesome

### Adding Doctypes

    var haml = require('haml')
    haml.doctypes.foo = '<!DOCTYPE ... >'
    
Will now allow you to:
    !!! foo
    
## License 

(The MIT License)

Copyright (c) 2011 TJ Holowaychuk &lt;tj@vision-media.ca&gt;

Permission is hereby granted, free of charge, to any person obtaining
a copy of this software and associated documentation files (the
'Software'), to deal in the Software without restriction, including
without limitation the rights to use, copy, modify, merge, publish,
distribute, sublicense, and/or sell copies of the Software, and to
permit persons to whom the Software is furnished to do so, subject to
the following conditions:

The above copyright notice and this permission notice shall be
included in all copies or substantial portions of the Software.

THE SOFTWARE IS PROVIDED 'AS IS', WITHOUT WARRANTY OF ANY KIND,
EXPRESS OR IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF
MERCHANTABILITY, FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT.
IN NO EVENT SHALL THE AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY
CLAIM, DAMAGES OR OTHER LIABILITY, WHETHER IN AN ACTION OF CONTRACT,
TORT OR OTHERWISE, ARISING FROM, OUT OF OR IN CONNECTION WITH THE
SOFTWARE OR THE USE OR OTHER DEALINGS IN THE SOFTWARE.

History.md

0.6.1 / 2012-07-26 
==================

  * fix global leak of Parser

0.6.0 / 2012-06-25 
==================

  * remove engine restriction
  * [fix] Changed require('sys') to require('util') for compatibility with node v0.8
  * require('util') instead of require('sys')
  * Refactor tests to not use require.paths
  * Typechange (from a symlink to a file).

0.5.1 / 2011-03-30 
==================

  * Exposing Parser for extension [Derek Hammer]

0.5.0 / 2011-03-04 
==================

  * Added `.compile()` for Express 2.0
  * Fixed issue with to many newlines in output. Closes #24

0.4.5 / 2010-06-04
==================

  * Ignoring stray indent

0.4.4 / 2010-05-27
==================

  * Fixed arbitrary whitespace support

0.4.3 / 2010-05-25
==================

  * Fixed support for CRLF / CR line-endings, converted on input
  * Exporting HamlError

0.4.2 / 2010-05-25
==================

  * Added HamlError
  * Buffer newline indentation. Closes #23
  * Benchmarks with node-bench

0.4.1 / 2010-05-17
==================

  * Fixed "- each" with non-enumerables, no longer throws exception
  * Fixed array iteration

0.4.0 / 2010-05-06
==================

  * Using Function constructor instead of eval()
  * Performance enhanced by faster iteration implementation for "- each"

0.3.1 / 2010-04-26
==================

  * Fixed support for tags with hypens (both namespace / tag name, ex: "fb:login-button")

0.3.0 / 2010-04-16
==================

  * Added xml namespace support
  * Added xml support

0.2.0 / 2010-04-06
==================

  * Added conditional comment support [ciaranj]
  * Fixed; Trimming input string before tokenization
  * Fixed issue requiring quoting of "for" when used in attrs [aheckmann]
  * Fixed; Exposing Haml compilation cache. Closes #12
  * Fixed :javascript "type" attr, now "text/javascript" [aheckmann]

0.1.0 / 2010-03-31
==================

  * Added "cache" option, making haml.js over 90 times faster than haml-js
  * Improved textBlock whitespace replication
  * Fixed empty tags followed by class / ids on new lines. Closes #6

0.0.4 / 2010-03-29
==================

  * Added better error reporting

0.0.3 / 2010-03-29
==================

  * Added "filename" option support to aid in error reporting
  * Added exports.compile() to create intermediate javascript
  * Added `make benchmark`
  * Changed; caching function templates to increase performance
  * Fixed; ids and classes allowing underscores. Closes #5
  * Fixed outdent issue when \n is not followed by whitespace. Closes #8

0.0.2 / 2010-03-26
==================

  * Added haml.js vs haml-js benchmarks
  * Fixed; commenting :javascript CDATA

0.0.1 / 2010-03-26
==================

  * Initial release


spec/node.js

// require.paths.unshift('spec', './spec/lib', 'lib')
haml = require('../lib/haml')
require('./lib/jspec')
require('./unit/spec.helper')

JSpec
  .exec('spec/unit/spec.js')
  .run({ reporter: JSpec.reporters.Terminal, fixturePath: 'spec/fixtures', failuresOnly: true })
  .report()


test/test.js
'use strict';

var fs = require('fs');
var haml = require('../lib/haml');

describe('haml', function () {
  describe('.version', function () {
    it('should be a triplet', function () {
      haml.version.should.match(/^\d+\.\d+\.\d+$/);
    });
  });

  describe('.compile()', function () {
    it('should return a function', function () {
      var fn = haml.compile('%foo= bar');
      fn({ bar: 'baz' }).should.equal('\n<foo>baz</foo>');
    });
  });

  describe('Parser', function () {
    it('should be available for extension', function () {
      haml.Parser.should.not.equal(undefined);
    });
  });

  describe('.render()', function () {
    var assertAs, assert, assertXML;

    // Prepare helper functions
    before(function () {
      assertAs = function(name, type, options) {
        var str = fs.readFileSync(__dirname + '/fixtures/' + name + '.haml').toString();
        try {
          var html = haml.render(str, options).trim();
          var expected = fs.readFileSync(__dirname + '/fixtures/' + name + '.' + type).toString().trim();
          html.should.equal(expected);
        } catch (err) {
          throw err;
        }
      };
      assert = function(name, options) {
        assertAs(name, 'html', options, 'CRLF', '\r\n');
      };
      assertXML = function(name, options) {
        assertAs(name, 'xml', options, 'CRLF', '\r\n');
      };
    });

    it('should allow passing of a context object', function () {
      assert('context', { context: 'yay' });
    });

    it('should allow passing of literals', function () {
      assert('literals', { locals: { user: 'tj' }});
    });

    it('should not fail on trailing indents', function () {
      assert('trailing-indent');
    });

    it('should add xml support via the "xml" option', function () {
      assertXML('feed', { xml: true });
    });

    it('should support xml namespaces', function () {
      assertXML('namespace');
    });

    it('should utilize "filename" option when an error is thrown', function () {
      try { assert('error', { filename: 'error.haml' }); }
      catch (err) {
        err.message.should.eql('(error.haml):3 invalid indentation; got 3, when previous was 1');
      }
    });

    it('should default filename to "Haml" when an error is thrown', function () {
      try { assert('error'); }
      catch (err) {
        err.message.should.eql('(Haml):3 invalid indentation; got 3, when previous was 1');
      }
    });

    it('should bitch when "cache" is true without a filename given', function () {
      // -{ assert('tag.simple', { cache: true }) }.should.throw_error
    });

    it('should pre-compiled and cache when "cache" is true', function () {
      assert('tag.simple', { cache: true, filename: 'tag.simple.haml' });
      assert('tag.simple', { cache: true, filename: 'tag.simple.haml' });
    });

    describe('\\n', function () {
      it('should support blank lines', function () {
        assert('newlines');
      });

      it('should support blank lines within tags', function () {
        assert('newlines.within-tags');
      });
    });

    describe('.class', function () {
      it('should output a div with the given class', function () {
        assert('class');
      });

      it('should work with several classes', function () {
        assert('classes');
      });
    });

    describe('#id', function () {
      it('should output a div with the given id', function () {
        assert('id');
      });
    });

    describe('%tag', function () {
      it('should work with no text or block', function () {
        assert('tag.simple');
      });

      it('should work with text', function () {
        assert('tag.text');
      });

      it('should work with block text', function () {
        assert('tag.text.block');
      });

      it('should work with blocks of text and tags', function () {
        assert('tag.text.block.complex');
      });

      it('should work with many classes / ids / attrs', function () {
        assert('tag.complex');
      });

      it('should allow empty tags', function () {
        assert('tag.empty');
      });
    });

    describe('%tag.class', function () {
      it('should output tag with a class', function () {
        assert('tag.class');
      });

      it('should work with several classes', function () {
        assert('tag.classes');
      });

      it('should support self-closing tags', function () {
        assert('tag.self-close');
      });
    });

    describe('%tag!=', function () {
      it('should output the evaluated code', function () {
        assert('tag.code');
      });

      it('should not escape output', function () {
        assert('tag.code.no-escape');
      });
    });

    describe('%tag=', function () {
      it('should escape the evaluated code', function () {
        assert('tag.escape');
      });
    });

    describe('%namespace:tag', function () {
      it('should output a tag with a namespace prefix', function () {
        assert('namespace.tag');
      });
    });

    describe('{...}', function () {
      it('should be mapped as html attributes', function () {
        assert('tag.attrs');
      });

      it('should escape values', function () {
        assert('tag.attrs.escape');
      });

      it('should allow booleans', function () {
        assert('tag.attrs.bools');
      });
    });

    describe('!!!', function () {
      it('should default the doctype to 1.0 transitional', function () {
        assert('doctype');
      });
    });

    describe('!!! NAME', function () {
      it('should output a specific doctype', function () {
        assert('doctype.xml');
      });

      it('should be case-insensitive', function () {
        assert('doctype.xml.case');
      });
    });

    describe('nesting', function () {
      it('should work when nested downwards', function () {
        assert('nesting.simple');
      });

      it('should work when blocks outdent', function () {
        assert('nesting.complex');
      });
    });

    describe('- code', function () {
      it('should work with if statements', function () {
        assert('code.if');
      });

      it('should work when nested', function () {
        assert('code.nested');
      });
    });

    describe('- each', function () {
      it('should iterate', function () {
        assert('code.each', { locals: { items: ['one', 'two', 'three'] }});
        assert('code.each.non-enumerable', { locals: { items: null }});
      });

      it('should iterate objects', function () {
        assert('code.each', { locals: { items: { 0: 'one', 1: 'two', 2: 'three' }}});
        assert('code.each.index', { locals: { items: { 0: 'one', 1: 'two', 2: 'three' }}});
      });

      it('should iterate with index', function () {
        assert('code.each.index', { locals: { items: ['one', 'two', 'three'] }});
      });
    });

    describe('= code', function () {
      it('should output evaluation', function () {
        assert('code');
      });
    });

    describe('&= code', function () {
      it('should output evaluation while escaping html entities', function () {
        assert('code.escape');
      });
    });

    describe('<literal></html>', function () {
      it('should remain intact', function () {
        assert('html');
      });
    });

    describe('\\char', function () {
      it('should escape the character', function () {
        assert('escape');
      });
    });

    describe('-#', function () {
      it('should become a silent comment', function () {
        assert('comment');
      });
    });

    describe('/', function () {
      it('should comment out tags', function () {
        assert('comment.tag');
      });

      it('should comment out blocks', function () {
        assert('comment.block');
      });

      it('should comment out text', function () {
        assert('comment.text');
      });

      it('should work in blocks', function () {
        assert('comment.text.complex');
      });
    });

    describe('/[]', function () {
      it('should insert conditional comment blocks', function () {
        assert('comment.block.conditional');
      });
    });

    describe(':filter', function () {
      describe('plain', function () {
        it('should ignore haml specific characters', function () {
          assert('filter.plain');
        });
      });

      describe('cdata', function () {
        it('should wrap with CDATA tags', function () {
          assert('filter.cdata');
        });

        it('should retain whitespace', function () {
          assert('filter.cdata.whitespace');
        });
      });

      describe('javascript', function () {
        it('should wrap with <script> and CDATA tags', function () {
          assert('filter.javascript');
        });
      });
    });

    describe('bug fixes', function () {
      it('#8 code block', function () {
        assert('issue.#8', { locals: { items: ['foo', 'bar', 'baz'] }});
      });

      it('#10 Attributes should not need quotes', function () {
        assert('issue.#10');
      });
    });
  });
});


benchmarks/run.js

var sys = require('sys'),
    fs = require('fs'),
    haml = require('../lib/haml'),
    hamlJS = require('./haml-js/lib/haml'),
    page = fs.readFileSync('benchmarks/page.haml')
    times = 3000

function bm(label, fn) {
  var start = +new Date
  fn()
  var duration = ((+new Date) - start) / 1000
  sys.puts(label + ': ' + duration + ' seconds')
}

sys.puts(times + ' times')

bm('null', function(){
  var n = times
  while (n--) ;
})

sys.puts('')

bm('haml.js', function(){
  var n = times
  while (n--) haml.render(page)
})

bm('haml.js cached', function(){
  var n = times
  while (n--) haml.render(page, { cache: true, filename: 'page.haml' })
})

sys.puts('')

bm('haml-js', function(){
  var n = times
  while (n--) hamlJS.render(page)
})

var js = hamlJS.compile(page)

bm('haml-js cached', function(){
  var n = times
  while (n--) hamlJS.execute(js)
})

var js = hamlJS.optimize(hamlJS.compile(page))

bm('haml-js cached / optimized', function(){
  var n = times
  while (n--) hamlJS.execute(js)
})


lib/haml.js
// Haml - Copyright TJ Holowaychuk <tj@vision-media.ca> (MIT Licensed)

/**
 * Module dependencies.
 */

var sys = require('util')

/**
 * Version.
 */

exports.version = '0.6.1'

/**
 * Haml template cache.
 */

exports.cache = {}

/**
 * Default error context length.
 */

exports.errorContextLength = 15

/**
 * Self closing tags.
 */
 
exports.selfClosing = [
    'meta',
    'img',
    'link',
    'br',
    'hr',
    'input',
    'area',
    'base'
  ]

/**
 * Default supported doctypes.
 */

exports.doctypes = {
  '5': '<!DOCTYPE html>',
  'xml': '<?xml version="1.0" encoding="utf-8" ?>',
  'default': '<!DOCTYPE html PUBLIC "-//W3C//DTD XHTML 1.0 Transitional//EN" "http://www.w3.org/TR/xhtml1/DTD/xhtml1-transitional.dtd">',
  'strict': '<!DOCTYPE html PUBLIC "-//W3C//DTD XHTML 1.0 Strict//EN" "http://www.w3.org/TR/xhtml1/DTD/xhtml1-strict.dtd">',
  'frameset': '<!DOCTYPE html PUBLIC "-//W3C//DTD XHTML 1.0 Frameset//EN" "http://www.w3.org/TR/xhtml1/DTD/xhtml1-frameset.dtd">',
  '1.1': '<!DOCTYPE html PUBLIC "-//W3C//DTD XHTML 1.1//EN" "http://www.w3.org/TR/xhtml11/DTD/xhtml11.dtd">',
  'basic': '<!DOCTYPE html PUBLIC "-//W3C//DTD XHTML Basic 1.1//EN" "http://www.w3.org/TR/xhtml-basic/xhtml-basic11.dtd">',
  'mobile': '<!DOCTYPE html PUBLIC "-//WAPFORUM//DTD XHTML Mobile 1.2//EN" "http://www.openmobilealliance.org/tech/DTD/xhtml-mobile12.dtd">'
}

/**
 * Default filters.
 */
 
exports.filters = {
  
  /**
   * Return plain string.
   */
  
  plain: function(str) {
    return str
  },
  
  /**
   * Wrap with CDATA tags.
   */
  
  cdata: function(str) {
    return '<![CDATA[\n' + str + '\n]]>'
  },
  
  /**
   * Wrap with <script> and CDATA tags.
   */
  
  javascript: function(str) {
    return '<script type="text/javascript">\n//<![CDATA[\n' + str + '\n//]]></script>'
  }
}

/**
 * Function templates.
 */
 
exports.templates = {
  
  /**
   * Execute __code__.
   */
  
  code: (function(){
    __code__;
    return ""
  }).toString(),
  
  /**
   * Execute __code__ followed by buffering __block__.
   */
  
  codeBlock: (function(){
    var buf = [];
    __code__
    buf.push(__block__);
    return buf.join("")
  }).toString(),
  
  /**
   * Iterate __vals__ as __val__ while buffering __block__.
   */
  
  iterate: (function(){
    var buf = [];
    if (__vals__ instanceof Array) {
      for (var i = 0, len = __vals__.length; i < len; ++i) {
        var __key__ = i;
        var __val__ = __vals__[i];
        buf.push(__block__);
      }
    } else if (__vals__) {
      var keys = Object.keys(__vals__);
      for (var i = 0, len = keys.length; i < len; ++i) {
        var __key__ = keys[i];
        var __val__ = __vals__[__key__];
        buf.push(__block__);
      }
    }
    return buf.join("")
  }).toString()
}

/**
 * HamlError.
 */

var HamlError = exports.HamlError = function(msg) {
    this.name = 'HamlError'
    this.message = msg
    Error.captureStackTrace(this, exports.render)
}
sys.inherits(HamlError, Error)

/**
 * Lexing rules.
 */

var rules = {
  indent: /^\n( *)(?! *-#)/,
  conditionalComment: /^\/(\[[^\n]+\])/,
  comment: /^\n? *\/ */,
  silentComment: /^\n? *-#([^\n]*)/,
  doctype: /^!!! *([^\n]*)/,
  escape: /^\\(.)/,
  filter: /^:(\w+) */,
  each: /^\- *each *(\w+)(?: *, *(\w+))? * in ([^\n]+)/,
  code: /^\-([^\n]+)/,
  outputCode: /^!=([^\n]+)/,
  escapeCode: /^=([^\n]+)/,
  attrs: /^\{(.*?)\}/,
  tag: /^%([-a-zA-Z][-a-zA-Z0-9:]*)/,
  class: /^\.([\w\-]+)/,
  id: /^\#([\w\-]+)/,
  text: /^([^\n]+)/
}

/**
 * Return error context _str_.
 *
 * @param  {string} str
 * @return {string}
 * @api private
 */

function context(str) {
  return String(str)
    .substr(0, exports.errorContextLength)
    .replace(/\n/g, '\\n')
}

/**
 * Tokenize _str_.
 *
 * @param  {string} str
 * @return {array}
 * @api private
 */

function tokenize(str) {
  var captures,
      token,
      tokens = [],
      line = 1,
      lastIndents = 0,
      str = String(str).trim().replace(/\r\n|\r/g, '\n')
  function error(msg){ throw new HamlError('(Haml):' + line + ' ' + msg) }
  while (str.length) {
    for (var type in rules)
      if (captures = rules[type].exec(str)) {
        token = {
          type: type,
          line: line,
          match: captures[0],
          val: captures.length > 2
            ? captures.slice(1)
            : captures[1]
        }
        str = str.substr(captures[0].length)
        if (type === 'indent') ++line
        else  break
        var indents = token.val.length / 2
        if (indents % 1)
          error('invalid indentation; got ' + token.val.length + ' spaces, should be multiple of 2')
        else if (indents - 1 > lastIndents)
          error('invalid indentation; got ' + indents + ', when previous was ' + lastIndents)
        else if (lastIndents > indents)
          while (lastIndents-- > indents)
            tokens.push({ type: 'outdent', line: line })
        else if (lastIndents !== indents)
          tokens.push({ type: 'indent', line: line })
        else
          tokens.push({ type: 'newline', line: line })
        lastIndents = indents
      }
    if (token) {
      if (token.type !== 'silentComment')
        tokens.push(token)
      token = null
    } else 
      error('near "' + context(str) + '"')
  }
  return tokens.concat({ type: 'eof' })
}

/**
 * Render template _name_ with the given placeholder _vals_.
 *
 * @param  {string} name
 * @param  {object} vals
 * @return {string}
 * @api private
 */

function template(name, vals) {
  var buf = '(' + exports.templates[name] + ').call(this)'
  for (var key in vals)
    buf = buf.replace(new RegExp(key, 'g'), vals[key])
  return buf
}

// --- Parser

/**
 * Initialize parser with _str_ and _options_.
 */

var Parser = exports.Parser = function (str, options) {
  options = options || {}
  this.tokens = tokenize(str)
  this.xml = options.xml
}

Parser.prototype = {
  
  /**
   * Lookahead a single token.
   *
   * @return {object}
   * @api private
   */
  
  get peek() {
    return this.tokens[0]
  },
  
  /**
   * Advance a single token.
   *
   * @return {object}
   * @api private
   */
  
  get advance() {
    return this.current = this.tokens.shift()
  },
  
  /**
   *    outdent
   *  | eof
   */
  
  get outdent() {
    switch (this.peek.type) {
      case 'eof':
        return
      case 'outdent':
        return this.advance
      default:
        throw new HamlError('expected outdent, got ' + this.peek.type)
    }
  },
  
  /**
   * text
   */
  
  get text() {
    return '"' + this.advance.val.trim() + '"'
  },
  
  /**
   * indent expr outdent
   */
  
  get block() {
    var buf = []
    this.advance
    while (this.peek.type !== 'outdent' &&
           this.peek.type !== 'eof')
      buf.push(this.expr)
    this.outdent
    return buf.join(' + ')
  },
  
  /**
   * indent expr
   */
  
  get textBlock() {
    var token,
        indents = 1,
        buf = []
    this.advance
    while (this.peek.type !== 'eof' && indents)
      switch((token = this.advance).type) {
        case 'newline':
          buf.push('"\\n' + Array(indents).join('  ') + '"')
          break
        case 'indent':
          ++indents
          buf.push('"\\n' + Array(indents).join('  ') + '"')
          break
        case 'outdent':
          --indents
          if (indents === 1) buf.push('"\\n"')
          break
        default:
          buf.push('"' + token.match.replace(/"/g, '\\\"') + '"')
      }
    return buf.join(' + ')
  },
  
  /**
   *  ( attrs | class | id )*
   */
  
  get attrs() {
    var attrs = ['attrs', 'class', 'id'],
        classes = [],
        buf = []
    while (attrs.indexOf(this.peek.type) !== -1)
      switch (this.peek.type) {
        case 'id':
          buf.push('id: "' + this.advance.val + '"')
          break
        case 'class':
          classes.push(this.advance.val)
          break
        case 'attrs':
          buf.push(this.advance.val.replace(/(for) *:/gi, '"$1":'))
      }
    if (classes.length)
      buf.push('"class": "' + classes.join(' ') + '"')
    return buf.length 
      ? ' " + attrs({' + buf.join() + '}) + "'
      : ''
  },
  
  /**
   *   tag
   * | tag text
   * | tag conditionalComment
   * | tag comment
   * | tag outputCode
   * | tag escapeCode
   * | tag block
   */
  
  get tag() {
    var tag = this.advance.val,
        selfClosing = !this.xml && exports.selfClosing.indexOf(tag) !== -1,
        buf = ['"\\n<' + tag + this.attrs + (selfClosing ? '/>"' : '>"')]
    switch (this.peek.type) {
      case 'text':
        buf.push(this.text)
        break
      case 'conditionalComment':
        buf.push(this.conditionalComment)
        break;
      case 'comment':
        buf.push(this.comment)
        break
      case 'outputCode':
        buf.push(this.outputCode)
        break
      case 'escapeCode':
        buf.push(this.escapeCode)
        break
      case 'indent':
        buf.push(this.block)
    }
    if (!selfClosing) buf.push('"</' + tag + '>"')
    return buf.join(' + ')
  },
  
  /**
   * outputCode
   */
  
  get outputCode() {
    return this.advance.val
  },
  
  /**
   * escapeCode
   */
  
  get escapeCode() {
    return 'escape(' + this.advance.val + ')'
  },
  
  /**
   * doctype
   */
  
  get doctype() {
    var doctype = this.advance.val.trim().toLowerCase() || 'default'
    if (doctype in exports.doctypes)
      return '"' + exports.doctypes[doctype].replace(/"/g, '\\"') + '"'
    else
      throw new HamlError("doctype `" + doctype + "' does not exist")
  },
  
  /**
   * conditional comment expr
   */

  get conditionalComment() {
    var condition= this.advance.val
    var buf = this.peek.type === 'indent'
      ? this.block
      : this.expr
    return '"<!--' + condition + '>" + (' + buf + ') + "<![endif]-->"'
  },

  /**
   * comment expr
   */
  
  get comment() {
    this.advance
    var buf = this.peek.type === 'indent'
      ? this.block
      : this.expr
    return '"<!-- " + (' + buf + ') + " -->"'
  },
  
  /**
   *   code
   * | code block 
   */
  
  get code() {
    var code = this.advance.val
    if (this.peek.type === 'indent')
      return template('codeBlock', { __code__: code, __block__: this.block })
    return template('code', { __code__: code })
  },
  
  /**
   * filter textBlock
   */
  
  get filter() {
    var filter = this.advance.val
    if (!(filter in exports.filters))
      throw new HamlError("filter `" + filter + "' does not exist")
    if (this.peek.type !== 'indent')
      throw new HamlError("filter `" + filter + "' expects a text block")
    return 'exports.filters.' + filter + '(' + this.textBlock + ')'
  },
  
  /**
   * each block
   */
  
  get iterate() {
    var each = this.advance
    if (this.peek.type !== 'indent')
      throw new HamlError("'- each' expects a block, but got " + this.peek.type)
    return template('iterate', {
      __key__: each.val[1],
      __vals__: each.val[2],
      __val__: each.val[0],
      __block__: this.block
    })
  },
  
  /**
   *   eof
   * | tag
   * | text*
   * | each
   * | code
   * | escape
   * | doctype
   * | filter
   * | comment
   * | conditionalComment
   * | escapeCode
   * | outputCode
   */
  
  get expr() {
    switch (this.peek.type) {
      case 'id': 
      case 'class':
        this.tokens.unshift({ type: 'tag', val: 'div' })
        return this.tag
      case 'tag':
        return this.tag
      case 'text':
        var buf = []
        while (this.peek.type === 'text') {
          buf.push(this.advance.val.trim())
          if (this.peek.type === 'newline')
            this.advance
        }
        return '"' + buf.join(' ') + '"'
      case 'each':
        return this.iterate
      case 'code':
        return this.code
      case 'escape':
        return '"' + this.advance.val + '"'
      case 'doctype':
        return this.doctype
      case 'filter':
        return this.filter
      case 'conditionalComment':
        return this.conditionalComment
      case 'comment':
        return this.comment
      case 'escapeCode':
        return this.escapeCode
      case 'outputCode':
        return this.outputCode
      case 'newline':
      case 'indent':
      case 'outdent':
        this.advance
        return this.expr
      default:
        throw new HamlError('unexpected ' + this.peek.type)
    }
  },
  
  /**
   * expr*
   */
  
  get js() {
    var buf = []
    while (this.peek.type !== 'eof')
      buf.push(this.expr)
    return buf.join(' + ')
  }
}

/**
 * Escape html entities in _str_.
 *
 * @param  {string} str
 * @return {string}
 * @api private
 */

function escape(str) {
  return String(str)
    .replace(/&/g, '&amp;')
    .replace(/>/g, '&gt;')
    .replace(/</g, '&lt;')
    .replace(/"/g, '&quot;')
}

/**
 * Render _attrs_ to html escaped attributes.
 *
 * @param  {object} attrs
 * @return {string}
 * @api public
 */

function attrs(attrs) {
  var buf = []
  for (var key in attrs)
    if (typeof attrs[key] === 'boolean') {
      if (attrs[key] === true)
        buf.push(key + '="' + key + '"')
    } else if (attrs[key])
      buf.push(key + '="' + escape(attrs[key]) + '"')
  return buf.join(' ')
}

/**
 * Compile a function from the given `str`.
 *
 * @param {String} str
 * @return {Function}
 * @api public
 */

exports.compile = function(str, options){
  var parser = new Parser(str, options);
  var fn = new Function('locals, attrs, escape, exports', 'with (locals || {}){ return ' + parser.js + '}');
  return function(locals){
    return fn.apply(this, [locals, attrs, escape, exports]);
  };
};

/**
 * Render a _str_ of haml.
 *
 * Options:
 *
 *   - locals   Local variables available to the template
 *   - context  Context in which the template is evaluated (becoming "this")
 *   - filename Filename used to aid in error reporting
 *   - cache    Cache compiled javascript, requires "filename"
 *   - xml      Force xml support (no self-closing tags)
 *
 * @param  {string} str
 * @param  {object} options
 * @return {string}
 * @api public
 */

exports.render = function(str, options) {
  var parser,
      options = options || {}
  if (options.cache && !options.filename)
    throw new Error('filename option must be passed when cache is enabled')
  return (function(){
    try {
      var fn
      if (options.cache && exports.cache[options.filename])
        fn = exports.cache[options.filename]
      else {
        parser = new Parser(str, options)
        fn = Function('locals, attrs, escape, exports', 'with (locals || {}){ return ' + parser.js + '}')
      }
      return (options.cache
          ? exports.cache[options.filename] = fn
          : fn).call(options.context, options.locals, attrs, escape, exports)
    } catch (err) {
      if (parser && err instanceof HamlError)
        err.message = '(Haml):' + parser.peek.line + ' ' + err.message
      else if (!(err instanceof HamlError))
        err.message = '(Haml): ' + err.message
      if (options.filename)
        err.message = err.message.replace('Haml', options.filename) 
      throw err
    }
  }).call(options.context)
}
